shared/utils/accessor-wrapper-types.hpp

```diff
@@ -37,54 +63,111 @@ namespace bs_hook {
     // We cannot EASILY solve this using wrapper types, because we could be holding a value type as T.
     // This makes things tricky and we should come back to this when we are confident we can handle this case correctly.
 
-
     template<internal::NTTPString name, class T, bool get, bool set>
     /// @brief Represents a InstanceProperty on a wrapper type. Forwards to calling the get/set methods where applicable.
     struct InstanceProperty;
     
     template<internal::NTTPString name, class T>
     struct InstanceProperty<name, T, true, false> {
-        explicit InstanceProperty(void* inst) noexcept : instance(inst) {}
+        explicit constexpr InstanceProperty(void* inst) noexcept : instance(inst) {}
         operator T() const {
             auto res = il2cpp_utils::GetPropertyValue<T, false>(reinterpret_cast<Il2CppObject*>(const_cast<void*>(instance)), name.data.data());
             if (!res) throw PropertyException(std::string("Failed to get instance property: ") + name.data.data());
             return *res;
         }
+
+        inline auto operator ->() const {
+            return this->operator T();
+        } 
+
+        inline auto operator *() const {
+            return this->operator T();
+        }
+
+        inline auto v() const {
+            return this->operator T();
+        }
+
+        template<typename... Targs>
+        decltype(auto) operator ()(Targs&&... args) { return this->operator T()(std::forward<Targs...>(args...)); }
+
         private:
         void* instance;
     };
 
     template<internal::NTTPString name, class T>
     struct InstanceProperty<name, T, false, true> {
-        explicit InstanceProperty(void* inst) noexcept : instance(inst) {}
+        explicit constexpr InstanceProperty(void* inst) noexcept : instance(inst) {}
         InstanceProperty& operator=(T&& t) {
             auto val = reinterpret_cast<Il2CppObject*>(instance);
             auto res = il2cpp_utils::SetPropertyValue<false>(val, name.data.data(), std::forward<decltype(t)>(t));
             if (!res) throw PropertyException(std::string("Failed to set instance property: ") + name.data.data());
             return *this;
         }
+        
         private:
         void* instance;
     };
 
+#define BINARY_OPERATOR_OP_EQ_PROP(op)      \
+template<typename U>                        \
+auto& operator op##=(const U& rhs) {        \
```

Comment:
Why do we force a materialization of `rhs` here? Can't we just forward this to the expression? What if the expression on T would use an rvalue reference? We would be losing out here by having this parameter be a `U const&`

Reply:
my answer to this is idk how to do that so I didn't code that

Reply:
Basically, does `template<class U>` and `U&& rhs` not work? This would allow for forwarding of moved parameters too

Reply:
no idea if that doesn't work